source/Concepts/Basic/About-Launch.rst

```diff
@@ -4,15 +4,33 @@ Launch
-The configuration of the system includes what programs to run, where to run them, what arguments to pass them, and ROS-specific conventions which make it easy to reuse components throughout the system by giving them each a different configuration.
-It is also responsible for monitoring the state of the processes launched, and reporting and/or reacting to changes in the state of those processes.
+A ROS 2 system typically consists of multiple nodes running across different processes (and even different machines).
+While it is possible to run each of these nodes separately, it can get cumbersome rather quickly.
 
-All of the above is specified in a launch file, which can be written in Python, XML, or YAML.
-This launch file can then be run using the ``ros2 launch`` command, and all of the nodes specified will be run.
+The purpose of the ``launch`` system in ROS 2 is to provide the means to orchestrate the execution of such a system. This tool uses _launch files_ to describe the system and its configuration, but not limited to which programs to run, how to run them and which arguments to use for them. These descriptions are then executed, starting processes, monitoring them, reporting on their state, and even reacting to their behavior.
 
-The `design document <https://design.ros2.org/articles/roslaunch.html>`__ details the goal of the design of ROS 2's launch system (not all functionality is currently available).
+System descriptions
+-------------------
+
+``launch`` works with _descriptions_, which, for the most part, amount to sequences of _actions_ executing in a _context_ or scope. All actions have side effects, which may be limited to the context in which they are running or extend to the larger execution environment (e.g. the local operating system). These side effects will typically outlive the context in which the corresponding action was run, which may be obvious for an action that changes an environment variable but not for an action that executes a process. Actions may also generate _events_, signals that are propagated through the context, that may be handled by _event handlers_.
+
+Actions alone allow for static descriptions that act upon and react to the environment in predefined ways. For a description to adapt to external input, _conditions_ and _substitutions_ exist. The former are boolean predicates, used for conditional actions, while the latter resemble text macros, useful for dynamic action configuration.
+
+In a way, ``launch`` descriptions are not that different from ASTs in procedural programming languages. The analogy brings about an important distinction that is easy to miss when writing launch files using the native Python APIs: neither actions nor conditions nor substitutions are computations but descriptions of computations to be carried out on launch. That is how actions can be implemented as sequences of other actions -- launch file inclusion being the paradigmatic example of this.
```

Comment:
Both good points.  Probably worth a link to: https://en.wikipedia.org/wiki/Abstract_syntax_tree

Reply:
> you are writing a description of a file it will generate.

It's not the description of a file, but a description of what `launch` should execute. Parsing XML/YAML yields the same object hierarchy. Taking the analogy further, start a Python shell and run:

```python
import ast as pyparser
code = "a = 1; print('a = ', a)"
ast = pyparser.parse(code)
print(pyparser.dump(ast, indent=4))
bytecode = compile(ast, "<string>", "exec")
exec(bytecode)
```

Oversimplifications aside, `ast` is the result of parsing Python `code`. We can then compile  `ast` into `bytecode`, which is executable. That's what the Python interpreter does when given a Python script.

`launch` does something similar. It takes a `LaunchDescription`, which may have been parsed from XML or YAML, or manually crafted like one does when writing launch files in Python, and instead of compiling anything, it executes it (again oversimplifying, actually the `LaunchDescription` gets traversed and executing `Action`s in it pushes coroutines to an event loop).

So when you are writing launch files in Python, it's as if you were writing the ASTs that Python works with or the intermediate representations that `gcc` compiles C++ to before generating machine code. That's why it may sometimes be difficult and painful (besides lack of documentation, that's also a thing).